process-engine/process_engine_core
# ♻️⚡️ Refactor FlowNodeHandling to Promise-Chains

**Requires:**
https://github.com/process-engine/process_engine_contracts/pull/100

**Changes:**

1. Remove all `execute` and `resume` logic from the `ExecuteProcessService` and `ResumeProcessService`.
    - The services no longer handle any kind of FlowNode execution.
    - The handlers now do this in a Promise chain, as described [here](https://github.com/process-engine/process_engine_core/issues/225).
2. Centralize the Subscription for `TerminateEndEvent` notifications to the `FlowNodeHandlerInterruptible` base handler.
    - This has the effect of providing all FlowNodeHandlers to subscribe for such a notification and react appropriately when it arrives.
3. Add `beforeExecute` and `afterExecute` hooks to the base FlowNodeHandler, which will allow derived handlers to perform customized setup- and cleanup- operations.
    - Handling subscriptions for TerminateEndEvents would be one application for these.
4. Remove the parent-handler pattern for `ServiceTasks`, `ParallelGateways`, `IntermediateCatchEvents` and `IntermediateThrowEvents`.
    - Dedicated FlowNodeHandlerFactories will now be used to create the various handler for those. 
5. Refactor all FlowNodeHandlers to return a list of `NextFlowNodes`, instead of just a single one. 
    - For most FlowNode types, only one outgoing SequenceFlow is still permitted.
    - Any type of Gateway can return as many SequenceFlows as they want (with the Exception of the ExclusiveGateway, which by design only returns one).
    - The ProcessModelFacade has been refactored to ensure that only Gateways will be allowed to have multiple `NextFlowNodes`.
    - This change makes it very easy to run parallel branches for a ParallelGateway.
6. Refactor ParallelGateways:
    - `ParallelSplitGateways` now do nothing but return a list of the FlowNodes starting each branch.
    - For each `ParallelJoinGateway`, a single instance is created and stored at the IoC Container.
    - The number of times such an instance can be called is equal to the incoming SequenceFlows attached to it.
    - Each time the instance is called, the result from the preceeding FlowNodeInstance is stored.
    - As soon as the expected number of results was received, the instance will resume the ProcessInstance with the next FlowNode and remove itself from the IoC container.
7. Implement `FlowNodePersistenceFacade`, to decouple the FlowNodeHandlers from the persistence logic.
8. Refactor BoundaryEvents to use a base `BoundaryEventHandler` class and decouple the handlers from their decorated handler.
    - In the old system, the BoundaryEvents executed the handler they were attached to. With the new chaining system, this chain has been reversed.
    - The BoundaryEvents get initialized by the handler they are attached to and trigger a callback each time their triggering event occurs.
    - The decorated handler also discards them, after it has finished execution.
    - This finally allows us to use non-interrupting BoundaryEvents and cyclic timers, because the decorated handlers can effectively run independently from their BoundaryEvents, but can also be interrupted by them, if need be.
9. Move creation of BoundaryEventHandlers into a dedicated factory.
10. Use customized error-type to handle errors thrown by an ErrorEndEvent (Bluebird doesn't event allow to reject non-errors).
11. Fix `cancelActivity` property parsing for BoundaryEvents
12. Cleanup `ProcessTokenFacade`:
    - Remove obsolete mapper functions
    - Expect a FlowNodeInstanceId to be passed to `addResultForFlowNode`
    - Add `containsResultForFlowNodeInstance` checker function

**Issues:**

Closes #212 
Closes #225
Part of #213
Part of https://github.com/process-engine/process_engine_runtime/issues/155
Part of https://github.com/process-engine/process_engine_runtime/issues/163
Part of https://github.com/process-engine/process_engine_runtime/issues/187
Part of https://github.com/process-engine/process_engine_runtime/issues/197

PR: #239

## How can others test the changes?

All in all, everything should run as before.
The only notable change should be in the ParallelGatewayHandlers, which should be a lot more reliable now.

## PR-Checklist

Please check the boxes in this list after submitting your PR:
